earthfile2llb/converter.go

```diff
@@ -2906,6 +2876,74 @@ func (c *Converter) newCmdID() int {
 	return cmdID
 }
 
+func (c *Converter) expandWildcardTargets(ctx context.Context, fullTargetName string) ([]string, error) {
+	parsedTarget, err := domain.ParseTarget(fullTargetName)
+	if err != nil {
+		return nil, err
+	}
+
+	if strings.Contains(fullTargetName, "**") {
+		return nil, errors.New("globstar (**) pattern not yet supported")
+	}
+
+	matches, err := c.opt.Resolver.ExpandWildcard(ctx, c.opt.GwClient, c.platr, c.target, parsedTarget)
+	if err != nil {
+		return nil, err
+	}
+
+	targets := make([]string, 0, len(matches))
+	for _, match := range matches {
+		childTargetName := fmt.Sprintf("./%s+%s", match, parsedTarget.GetName())
+
+		childTarget, err := domain.ParseTarget(childTargetName)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to parse target %q", childTargetName)
+		}
+
+		data, _, _, err := c.ResolveReference(ctx, childTarget)
+		if err != nil {
+			notExist := buildcontext.ErrEarthfileNotExist{}
+			if errors.As(err, &notExist) {
+				continue
+			}
+			return nil, errors.Wrapf(err, "unable to resolve target %q", childTargetName)
+		}
+
+		var found bool
+		for _, target := range data.Earthfile.Targets {
+			if target.Name == childTarget.GetName() {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			continue
+		}
+
+		targets = append(targets, childTargetName)
+	}
+
+	if len(targets) == 0 {
+		return nil, errors.Errorf("no matching targets found for pattern %q", parsedTarget.GetLocalPath())
+	}
+
+	return targets, nil
+}
+
+func clonesWithExpandedTargets[T cloneable[T]](expandedTargets []string, c T, setTarget func(t *T, expandedTarget string) error) ([]T, error) {
```

Comment:
Up to you, but I wonder if the use of generics here is more complicated than the simple loop plus `Clone` calls they are meant to abstract.

Reply:
yeah, I was wondering if the indirection is worth it and I'm open to removing it.
I tended to go with generics after all because I liked how it forces the user to clone the object, which I think it might be something that can be easily forgotten. 
I also considered whether this function can be used again, and figured it _might_ in something like `WITH DOCKER --load ./services/*+image`

Reply:
That sounds fair.